File: substrate/antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_compose.py
"""Antiek-bench weekly learn over recursive twin presentation write collective (pure).

backlog_mutated / store_mutated / suite_rewritten always False.
twin_written / merge_executed / draft_written always False.
live_router_authorized / secrets_stored always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.antiek_bench_weekly_usage_learn_compose import (
    AntiekBenchWeeklyUsageLearnCompose,
    AntiekBenchWeeklyUsageLearnComposeError,
    compose_antiek_bench_weekly_usage_learn,
)
from substrate.recursive_twin_presentation_write_twin_collective_fullscreen_mo_nd_twin_compose import (
    RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinCompose,
    RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinComposeError,
    compose_recursive_twin_presentation_write_twin_collective_fullscreen_mo_nd_twin,
)


class AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
    ValueError
):
    """Fail-closed validation for weekly learn + twin presentation write collective."""


@dataclass(frozen=True)
class AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose:
    week_id: str
    session_id: str
    parent_asset_id: str
    asset_id: str
    title: str
    account_id: str
    weekly_learn: AntiekBenchWeeklyUsageLearnCompose
    twin_presentation_pack: RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinCompose
    pack_ready: bool
    learn_ready: bool
    backlog_mutated: bool
    store_mutated: bool
    suite_rewritten: bool
    twin_written: bool
    prompts_injected: bool
    merge_executed: bool
    draft_written: bool
    analysis_written: bool
    live_dispatched: bool
    pack_dispatched: bool
    live_execution_authorized: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    remote_index_queried: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    inventory_mutated: bool
    charge_executed: bool
    record_persisted: bool
    purchase_executed: bool
    hosted: bool
    remote_fetched: bool
    live_dispatch_authorized: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "week_id": self.week_id,
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "asset_id": self.asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "weekly_learn": self.weekly_learn.to_dict(),
            "twin_presentation_pack": self.twin_presentation_pack.to_dict(),
            "pack_ready": self.pack_ready,
            "learn_ready": self.learn_ready,
            "backlog_mutated": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "twin_written": False,
            "prompts_injected": False,
            "merge_executed": False,
            "draft_written": False,
            "analysis_written": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "live_execution_authorized": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "remote_index_queried": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "inventory_mutated": False,
            "charge_executed": False,
            "record_persisted": False,
            "purchase_executed": False,
            "hosted": False,
            "remote_fetched": False,
            "live_dispatch_authorized": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_antiek_bench_weekly_learn_recursive_twin_presentation_write_collective(
    *,
    weekly_learn: object,
    twin_presentation_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose:
    """Weekly bench learn + recursive twin presentation write collective. Never mutates bench."""
    if not isinstance(operator_ack, bool):
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(weekly_learn, dict):
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            "weekly_learn must be an object"
        )
    if not isinstance(twin_presentation_pack, dict):
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            "twin_presentation_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "backlog_mutated=false · store_mutated=false · suite_rewritten=false",
        "twin_written=false · merge_executed=false · draft_written=false",
        "live_router_authorized=false · secrets_stored=false",
        "production_router_verdict=REJECT",
    ]

    try:
        wl = compose_antiek_bench_weekly_usage_learn(
            week_id=weekly_learn.get("week_id"),
            events=weekly_learn.get("events"),
            operator_ack=operator_ack,
            min_events_per_task=weekly_learn.get("min_events_per_task"),
        )
    except AntiekBenchWeeklyUsageLearnComposeError as e:
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            str(e)
        ) from e
    notes.extend(f"[weekly_learn] {n}" for n in wl.notes)

    try:
        tp = compose_recursive_twin_presentation_write_twin_collective_fullscreen_mo_nd_twin(
            twin=twin_presentation_pack.get("twin"),
            presentation=twin_presentation_pack.get("presentation"),
            write_pack=twin_presentation_pack.get("write_pack"),
            operator_ack=operator_ack,
            require_both=twin_presentation_pack.get("require_both"),
        )
    except RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinComposeError as e:
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            str(e)
        ) from e
    notes.extend(f"[twin_presentation_pack] {n}" for n in tp.notes)

    week = _require_nonempty(wl.week_id, field="week_id")
    session = _require_nonempty(tp.session_id, field="session_id")
    parent = _require_nonempty(tp.parent_asset_id, field="parent_asset_id")
    asset = _require_nonempty(tp.asset_id, field="asset_id")
    title = _require_nonempty(tp.title, field="title")
    account = _require_nonempty(tp.account_id, field="account_id")

    if require:
        pack_ready = (
            wl.learn_ready is True
            and tp.pack_ready is True
            and wl.backlog_mutated is False
            and wl.store_mutated is False
            and tp.twin_written is False
            and tp.merge_executed is False
            and tp.draft_written is False
            and tp.analysis_written is False
            and tp.live_dispatched is False
            and tp.live_execution_authorized is False
            and tp.live_router_authorized is False
            and tp.secrets_stored is False
            and tp.remote_index_queried is False
            and tp.pdf_primary is False
            and tp.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            operator_ack is True
            and wl.backlog_mutated is False
            and wl.store_mutated is False
            and tp.production_router_verdict == "REJECT"
            and tp.pdf_primary is False
            and tp.twin_written is False
            and (wl.learn_ready is True or tp.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — weekly bench learn + recursive twin presentation write "
            "collective ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — weekly_learn, twin_presentation_pack, or operator_ack gate open"
        )

    if (
        wl.backlog_mutated is not False
        or wl.store_mutated is not False
        or tp.twin_written is not False
        or tp.merge_executed is not False
        or tp.draft_written is not False
        or tp.analysis_written is not False
        or tp.live_dispatched is not False
        or tp.live_execution_authorized is not False
        or tp.live_router_authorized is not False
        or tp.secrets_stored is not False
        or tp.remote_index_queried is not False
        or tp.pdf_primary is not False
        or tp.production_router_verdict != "REJECT"
    ):
        raise AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "backlog_mutated=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "twin_written=false",
            "prompts_injected=false",
            "merge_executed=false",
            "draft_written=false",
            "analysis_written=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "live_execution_authorized=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "remote_index_queried=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "inventory_mutated=false",
            "charge_executed=false",
            "record_persisted=false",
            "purchase_executed=false",
            "hosted=false",
            "remote_fetched=false",
            "live_dispatch_authorized=false",
            "production_router_verdict=REJECT",
        )
    )

    return AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose(
        week_id=week,
        session_id=session,
        parent_asset_id=parent,
        asset_id=asset,
        title=title,
        account_id=account,
        weekly_learn=wl,
        twin_presentation_pack=tp,
        pack_ready=pack_ready,
        learn_ready=wl.learn_ready,
        backlog_mutated=False,
        store_mutated=False,
        suite_rewritten=False,
        twin_written=False,
        prompts_injected=False,
        merge_executed=False,
        draft_written=False,
        analysis_written=False,
        live_dispatched=False,
        pack_dispatched=False,
        live_execution_authorized=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        remote_index_queried=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        inventory_mutated=False,
        charge_executed=False,
        record_persisted=False,
        purchase_executed=False,
        hosted=False,
        remote_fetched=False,
        live_dispatch_authorized=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority=(
            "antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_compose_advisory"
        ),
    )


def format_antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_summary(
    c: AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"learn_ready={c.learn_ready} · "
        f"proposals={c.weekly_learn.proposal_count} · "
        f"twin_presentation_ready={c.twin_presentation_pack.pack_ready} · "
        f"week={c.week_id} · "
        f"verdict={c.production_router_verdict} · "
        "backlog_mutated=false · twin_written=false · suite_rewritten=false"
    )


__all__ = [
    "AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose",
    "AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveComposeError",
    "compose_antiek_bench_weekly_learn_recursive_twin_presentation_write_collective",
    "format_antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_summary",
]

File: apps/reading/src/api/antiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose.ts
/**
 * Antiek-bench weekly usage-learn residual over recursive twin presentation +
 * write twin collective fullscreen Midnight Oil unattended ND twin pack (pure).
 *
 * Operator vision: surface weekly bench learn (what worked / failed → rewrite
 * proposals) beside recursive twin presentation + write collective + fullscreen
 * + MO unattended + draft multiselect ND twin so settings / decision-tree
 * consumers see learn readiness with the full honesty stack — without mutating
 * bench backlog or store.
 *
 * backlog_mutated / store_mutated / suite_rewritten always false.
 * twin_written / merge_executed / draft_written always false.
 * live_router_authorized / secrets_stored always false.
 * production_router_verdict always REJECT.
 */

import {
  composeAntiekBenchWeeklyUsageLearn,
  type AntiekBenchWeeklyUsageLearnCompose,
  type AntiekBenchWeeklyUsageLearnInput,
} from "./antiekBenchWeeklyUsageLearnCompose";
import {
  composeRecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwin,
  type RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinCompose,
  type RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinInput,
} from "./recursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinCompose";

export interface AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveInput {
  weekly_learn: Omit<AntiekBenchWeeklyUsageLearnInput, "operator_ack">;
  twin_presentation_pack: Omit<
    RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose {
  week_id: string;
  session_id: string;
  parent_asset_id: string;
  asset_id: string;
  title: string;
  account_id: string;
  weekly_learn: AntiekBenchWeeklyUsageLearnCompose;
  twin_presentation_pack: RecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwinCompose;
  pack_ready: boolean;
  learn_ready: boolean;
  backlog_mutated: false;
  store_mutated: false;
  suite_rewritten: false;
  twin_written: false;
  prompts_injected: false;
  merge_executed: false;
  draft_written: false;
  analysis_written: false;
  live_dispatched: false;
  pack_dispatched: false;
  live_execution_authorized: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  remote_index_queried: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  inventory_mutated: false;
  charge_executed: false;
  record_persisted: false;
  purchase_executed: false;
  hosted: false;
  remote_fetched: false;
  live_dispatch_authorized: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Weekly bench learn stacked on recursive twin presentation write collective pack.
 * Never mutates bench; never writes twin; ND REJECT.
 */
export function composeAntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollective(
  input: AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveInput,
): AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.weekly_learn || typeof input.weekly_learn !== "object") {
    throw new Error("weekly_learn must be an object");
  }
  if (
    !input.twin_presentation_pack ||
    typeof input.twin_presentation_pack !== "object"
  ) {
    throw new Error("twin_presentation_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "backlog_mutated=false · store_mutated=false · suite_rewritten=false",
    "twin_written=false · merge_executed=false · draft_written=false",
    "live_router_authorized=false · secrets_stored=false",
    "production_router_verdict=REJECT",
  ];

  const weekly_learn = composeAntiekBenchWeeklyUsageLearn({
    ...input.weekly_learn,
    operator_ack: input.operator_ack,
  });
  notes.push(...weekly_learn.notes.map((n) => `[weekly_learn] ${n}`));

  const twin_presentation_pack =
    composeRecursiveTwinPresentationWriteTwinCollectiveFullscreenMoNdTwin({
      ...input.twin_presentation_pack,
      operator_ack: input.operator_ack,
    });
  notes.push(
    ...twin_presentation_pack.notes.map((n) => `[twin_presentation_pack] ${n}`),
  );

  const week_id = requireNonEmpty(weekly_learn.week_id, "week_id");
  const session_id = requireNonEmpty(
    twin_presentation_pack.session_id,
    "session_id",
  );
  const parent_asset_id = requireNonEmpty(
    twin_presentation_pack.parent_asset_id,
    "parent_asset_id",
  );
  const asset_id = requireNonEmpty(
    twin_presentation_pack.asset_id,
    "asset_id",
  );
  const title = requireNonEmpty(twin_presentation_pack.title, "title");
  const account_id = requireNonEmpty(
    twin_presentation_pack.account_id,
    "account_id",
  );

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      weekly_learn.learn_ready === true &&
      twin_presentation_pack.pack_ready === true &&
      weekly_learn.backlog_mutated === false &&
      weekly_learn.store_mutated === false &&
      twin_presentation_pack.twin_written === false &&
      twin_presentation_pack.merge_executed === false &&
      twin_presentation_pack.draft_written === false &&
      twin_presentation_pack.analysis_written === false &&
      twin_presentation_pack.live_dispatched === false &&
      twin_presentation_pack.live_execution_authorized === false &&
      twin_presentation_pack.live_router_authorized === false &&
      twin_presentation_pack.secrets_stored === false &&
      twin_presentation_pack.remote_index_queried === false &&
      twin_presentation_pack.pdf_primary === false &&
      twin_presentation_pack.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      weekly_learn.backlog_mutated === false &&
      weekly_learn.store_mutated === false &&
      twin_presentation_pack.production_router_verdict === "REJECT" &&
      twin_presentation_pack.pdf_primary === false &&
      twin_presentation_pack.twin_written === false &&
      (weekly_learn.learn_ready === true ||
        twin_presentation_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — weekly bench learn + recursive twin presentation write collective ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — weekly_learn, twin_presentation_pack, or operator_ack gate open",
    );
  }

  if (
    weekly_learn.backlog_mutated !== false ||
    weekly_learn.store_mutated !== false ||
    twin_presentation_pack.twin_written !== false ||
    twin_presentation_pack.merge_executed !== false ||
    twin_presentation_pack.draft_written !== false ||
    twin_presentation_pack.analysis_written !== false ||
    twin_presentation_pack.live_dispatched !== false ||
    twin_presentation_pack.live_execution_authorized !== false ||
    twin_presentation_pack.live_router_authorized !== false ||
    twin_presentation_pack.secrets_stored !== false ||
    twin_presentation_pack.remote_index_queried !== false ||
    twin_presentation_pack.pdf_primary !== false ||
    twin_presentation_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("merge_executed=false");
  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("live_execution_authorized=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("remote_index_queried=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("inventory_mutated=false");
  notes.push("charge_executed=false");
  notes.push("record_persisted=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("remote_fetched=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("production_router_verdict=REJECT");

  return {
    week_id,
    session_id,
    parent_asset_id,
    asset_id,
    title,
    account_id,
    weekly_learn,
    twin_presentation_pack,
    pack_ready,
    learn_ready: weekly_learn.learn_ready,
    backlog_mutated: false,
    store_mutated: false,
    suite_rewritten: false,
    twin_written: false,
    prompts_injected: false,
    merge_executed: false,
    draft_written: false,
    analysis_written: false,
    live_dispatched: false,
    pack_dispatched: false,
    live_execution_authorized: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    remote_index_queried: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    inventory_mutated: false,
    charge_executed: false,
    record_persisted: false,
    purchase_executed: false,
    hosted: false,
    remote_fetched: false,
    live_dispatch_authorized: false,
    production_router_verdict: "REJECT",
    notes,
    authority:
      "antiek_bench_weekly_learn_recursive_twin_presentation_write_collective_compose_advisory",
  };
}

export function formatAntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveSummary(
  c: AntiekBenchWeeklyLearnRecursiveTwinPresentationWriteCollectiveCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `learn_ready=${c.learn_ready} · ` +
    `proposals=${c.weekly_learn.proposal_count} · ` +
    `twin_presentation_ready=${c.twin_presentation_pack.pack_ready} · ` +
    `week=${c.week_id} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `backlog_mutated=false · twin_written=false · suite_rewritten=false`
  );
}
